### PR TITLE
helm: Fix duplicate tunnel-protocol configuration in cilium-config co…

### LINKED ISSUE #33456: Helm can't enable geneve protocol:  line 132: mapping key "tunnel-protocol" already defined at line 131
We are attempting to change our helm config to use Geneve instead of VXLAN, so what we can use DSR in our cluster.  Unfortunately, native routing mode is not an option in our case.

When attempting to deploy the helm chart with `tunnelProtocol: geneve` we get the error:
```
line 132: mapping key "tunnel-protocol" already defined at line 131
```

I believe this is due to the fact that the line referenced below is creating the tunnel-protocol: vxlan line.

https://github.com/cilium/cilium/blob/5614531067a83e20d24bccc7b12b314330d043c3/install/kubernetes/cilium/templates/cilium-configmap.yaml#L477

Then when it gets further down and tries to insert the geneve line, this is seen as a duplicate entry.

https://github.com/cilium/cilium/blob/5614531067a83e20d24bccc7b12b314330d043c3/install/kubernetes/cilium/templates/cilium-configmap.yaml#L481

I think the "default" in helm just needs to be a little smarter here to allow for properly enabling Geneve.

We are specifically running helm chart version 1.15.6, but it looks like this issue exists in the main branch and probably also affects 1.16.x release.

### PR DESCRIPTION
**Pull request by tylerauerbeck**
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [X] Thanks for contributing!

This updates updates the `cilium-config` configmap template to fix the potential for duplicate `tunnel-protocol` entries.

This is accomplished by removing the `tunnel-protocol` logic in earlier parts of the template (as they default to `vxlan`, which is considered to be the default setting in general). It then explicitly sets `vxlan` as the default value for `tunnel-protocol` in `values.yaml`.

By doing the above, it now results in only a single entry for `tunnel-protocol` and adjusts the values if it is set to something other than `vxlan` (i.e. `geneve`).

Fixes: #33456

```release-note
Simplify tunnel-protocol logic in helm templates to remove occurrence of duplicate entries. VXLAN is now explicitly set as the default value for `tunnel-protocol`.
```

### CODE REVIEW

**Comment by tylerauerbeck** [comment]
The only question I wanted to clarify is whether `aksbyocni` only supports `vxlan` or if it supports any of the protocols.

**Comment by squeed** [comment]
@tylerauerbeck to answer your immediate question, `aksbyocni` supports all tunneling protocols IIRC. So that's not a problem.

There are other cases that this change breaks, however:

1. `Values.gke.enabled` -- this requires `routing-mode: native` and no `tunnel-protocol`
2. `Values.routingMode: native` -- tunnelProtocol should not be set 

I think the right way to fix this is, unfortunately, to use a Helm variable as with other defaults up top.

**Comment by tylerauerbeck** [comment]
@squeed Went with the helm var approach. Let me know if this is what you had in mind.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 1]
```diff
@@ -1,6 +1,9 @@
+
```
Comment (squeed) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: nit: extra whitespace :-)

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 2]
```diff
@@ -84,6 +87,12 @@
   {{- $defaultK8sClientQPS = 10 -}}
   {{- $defaultK8sClientBurst = 20 -}}
 {{- end -}}
+
+{{- if or (.Values.gke.enabled) (eq .Values.routingMode "native") -}}
+  {{- $tunnelProto = "" -}}
+{{- else -}}
```
Comment (squeed) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: The logic should be something like:

```
if (gke || routingMode == native) {
    tunnelProto = ""
} else if (values.tunnelProto != "" {
    tunnelProto = values.tunnelProto
} else {
    tunnelProto = vxlan
}
```

Make sense?

Comment (tylerauerbeck) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: Updated to match the above.

**Review by squeed — APPROVED** [review]

**Comment by tylerauerbeck** [comment]
@squeed Sorry for the delay on this! And no problems, only one way to kill spiders 🔥 😬 

Let me fiddle with this a bit. My feeling is that `vxlan` seemed to be the default setting that appeared when another value wasn't set prior to these changes, so maybe it's okay to replicate that behavior here.

**Comment by ldelossa** [comment]
@tylerauerbeck - please be aware that your commit message structure is flagging the checkpatch linter. Check out the failing `BPF Checks / Check Patch` test for details.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 3]
```diff
@@ -84,6 +85,14 @@
...
   {{- $defaultK8sClientBurst = 20 -}}
 {{- end -}}
+
+{{- if or (.Values.gke.enabled) (eq .Values.routingMode "native") -}}
+  {{- $tunnelProto = "" -}}
+{{- else if ne (.Values.tunnelProtocol) "" -}}
+  {{- $tunnelProto = .Values.tunnelProtocol -}}
+{{- else -}}
+  {{- $tunnelProto = "vxlan" -}}
+{{- end -}}
```
Comment (giorio94) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:95]: A couple of drive-by comments:
* the tunnel protocol shall be configurable also in native routing mode. Indeed, there are certain features (e.g., Egress Gateway, DSR) which still create a tunnel for certain traffic categories in that case.
* the `tunnelProto` variable is currently never used, hence all this logic is not taking effect.

**Review by joestringer — CHANGES_REQUESTED** [review]
It looks like this is currently causing failures in common test cases per the CI results, and I see also that the feedback from @giorio94 needs to be addressed (`$tunnelProto` is unused; tunnel protocol must be configurable even if the routing mode is native, because some specific features may use tunnels even if regular pod<->pod traffic does not).

Several of the testing workflows are failing at "Wait for Cilium to become Ready". This likely indicates a bug with this PR in a particular configuration mode for Cilium. The workflow output has some indications of these failures with fatal error logs from the agent, and there should also be a sysdump attached to those workflows which you can open to look at the full agent logs from the test scenario. It's likely that there's just one particular case that needs to be fixed for this PR to pass most of these tests. If you're struggling with how to understand the testing results or investigate/reproduce, feel free to ask for further pointers here or in Slack.

The individual workflow failures should have a "sysdump" attached to it, which will include lots of information about the failure including logs from the cilium-agent which may point towards the underlying problem. But that said, I think that even if you start 

Marking "Request Changes" for now to make it clear that the PR must change to resolve the outstanding feedback before we merge the PR. The idea itself seems reasonable enough, it's just that the implementation breaks some common configurations.
